File: 21-Memento/src/go/editor_memento.go
package main

// EditorMemento is the Memento — an immutable snapshot. Its state field is
// unexported, so the caretaker can hold it but cannot read the content.
// Only code in this package (the originator) reads `content`.
type EditorMemento struct {
	content string
	label   string
}

// Label is safe metadata the caretaker may display without seeing the state.
func (m *EditorMemento) Label() string {
	return m.label
}

File: 21-Memento/src/rust/editor_memento.rs
/// Memento — an immutable snapshot. `content` is private; the originator
/// reads it via the crate-visible `content()`, while the caretaker only sees
/// the public `label()`. By discipline, the caretaker never calls `content()`.
pub struct EditorMemento {
    content: String,
    label: String,
}

impl EditorMemento {
    pub(crate) fn new(content: String, label: String) -> Self {
        EditorMemento { content, label }
    }

    // Read by the Originator to restore state.
    pub(crate) fn content(&self) -> &str {
        &self.content
    }

    // Safe metadata for the caretaker to display, without exposing the state.
    pub fn label(&self) -> &str {
        &self.label
    }
}
